metadata-ingestion/src/datahub/ingestion/source/sql/snowflake.py

```diff
@@ -214,34 +214,137 @@ def get_identifier(self, *, schema: str, entity: str, **kwargs: Any) -> str:
         regular = super().get_identifier(schema=schema, entity=entity, **kwargs)
         return f"{self.current_database.lower()}.{regular}"
 
+    def _populate_view_lineage(self) -> None:
+        # NOTE: This single query captures both the upstream and downstream table lineage for views.
+        # See https://docs.snowflake.com/en/sql-reference/account-usage/access_history.html#usage-notes for current limitations on capturing the lineage for views.
+        # Eg: For viewA->viewB->ViewC->TableD, snowflake does not yet log intermediate view logs, resulting in only the viewA->TableD edge.
+        view_lineage_query: str = """
+WITH view_lineage_history AS (
+  SELECT
+    vu.value : "objectName" AS view_name,
+    vu.value : "objectDomain" AS view_domain,
+    vu.value : "columns" AS view_columns,
+    r.value : "objectName" AS upstream_table_name,
+    r.value : "objectDomain" AS upstream_table_domain,
+    r.value : "columns" AS upstream_table_columns,
+    w.value : "objectName" AS downstream_table_name,
+    w.value : "objectDomain" AS downstream_table_domain,
+    w.value : "columns" AS downstream_table_columns,
+    t.query_start_time AS query_start_time
+  FROM
+    (
+      SELECT
+        *
+      FROM
+        snowflake.account_usage.access_history
+    ) t,
+    lateral flatten(input => t.DIRECT_OBJECTS_ACCESSED) vu,
+    lateral flatten(input => t.BASE_OBJECTS_ACCESSED) r,
+    lateral flatten(input => t.OBJECTS_MODIFIED) w
+  WHERE
+    vu.value : "objectId" IS NOT NULL
+    AND r.value : "objectId" IS NOT NULL
+    AND w.value : "objectId" IS NOT NULL
+    AND t.query_start_time >= to_timestamp_ltz({start_time_millis}, 3)
+    AND t.query_start_time < to_timestamp_ltz({end_time_millis}, 3)
+)
+SELECT
+  view_name,
+  view_columns,
+  upstream_table_name,
+  upstream_table_domain,
+  upstream_table_columns,
+  downstream_table_name,
+  downstream_table_domain,
+  downstream_table_columns
+FROM
+  view_lineage_history
+WHERE
+  view_domain in ('View', 'Materialized view')
+  AND view_name != upstream_table_name
+  AND upstream_table_name != downstream_table_name
+  AND view_name != downstream_table_name
+  QUALIFY ROW_NUMBER() OVER (
+    PARTITION BY view_name,
+    upstream_table_name,
+    downstream_table_name
+    ORDER BY
+      query_start_time DESC
+  ) = 1
+        """.format(
+            start_time_millis=int(self.config.start_time.timestamp() * 1000),
```

Comment:
So this means lineage will only be extracted for views which have been accessed in the time window?

Reply:
That's always true for lineage, because we depend on access history.